### ISSUE TITLE
Listening stops with many connections (pub/sub)

### ISSUE THREAD

**Issue by highfield** [nats-io/nats.net.v1#190]
My demo bench is a small Net Core app with 20 connections as publishers, and other 20 as listeners. The broker is reached as 127.0.0.1 on Windows 10.
Still I haven't understand what's going on, but I found an issue at this point:
https://github.com/nats-io/csharp-nats/blob/d2ed82da04b2d2ddee770f1424302c0faf482289/NATS.Client/Conn.cs#L1047
Basically, the task does not start although apparently there's no reason on that. The result is that the inner method is not called, and _no_ message is listen to any of the subscribers.
I solved turning the line as:
`            new Thread(() => spinUpSocketWatchers()).Start();
`
Another way is to calling the method directly, but I guess you don't want to halt the caller.
I'll dig into a little more, but...BAH!...let me know your opinion on.

**Comment by ColinSullivan1** [comment]
@highfield , This is strange, especially with only twenty subscribers.  Is your application under heavy load?  If the publishers are going full blast, it may be.  Under heavy load, I've seen tasks take awhile to start and behave strangely in other ways.  Threads are given more resources and priority in .NET, which might explain why a thread would work, and a task wouldn't.

I wonder if reducing the # of tasks might help here.  Can you set the [SubscriberDeliveryTaskCount](http://nats-io.github.io/csharp-nats/class_n_a_t_s_1_1_client_1_1_options.html#a030dfa5eac04072735d7529147d5c562) option to see this could be case?  There is some example code [here](https://github.com/nats-io/csharp-nats/blob/v0.8.0/NATSUnitTests/UnitTestSub.cs#L738).

Thanks,
Colin

**Comment by highfield** [comment]
Well, I'm still digging the code (I grabbed the latest one), but there are more than just one thing strange.

The most used pattern is to create many connections from within the same thread. That works, although it's VERY slow to start. The below code takes over 2 minutes just to create the connections, then the loop starts. I wonder why it takes so long, because a straight-raw connection of the same quantity of TcpClient-s takes way less than a single second (against the same broker, of course).
```
    class NatsOrigSingleThreadDemo
    {
        const int N = 1 * 1000;

        string[] servers = new string[]
        {
            //"nats://127.0.0.1:4222",
            "nats://localhost:4222"
        };


        public void Run()
        {
            this.SingleThread();
        }


        private void SingleThread()
        {
            Options opts = ConnectionFactory.GetDefaultOptions();
            opts.PingInterval = 30000;
            opts.Timeout = 3000;
            opts.Servers = servers;

            ConnectionFactory cf = new ConnectionFactory();

            IConnection[] subscribers = null;
            IConnection[] publishers = null;
            try
            {
                subscribers = Enumerable.Range(0, 100)
                    .Select(_ => cf.CreateConnection(opts))
                    .ToArray();

                publishers = Enumerable.Range(0, 100)
                    .Select(_ => cf.CreateConnection(opts))
                    .ToArray();

                Thread.Sleep(5000);
                Console.WriteLine("starting pubblication...");

                for (int i = 0; i < N; i++)
                {
                    var c = publishers[i % 100];
                    var jreq = new JObject();
                    jreq["n"] = i;
                    jreq["descr"] = "Creates a live connection";
                    jreq["dt"] = DateTime.Now;
                    byte[] buffer = Encoding.UTF8.GetBytes(jreq.ToString());
                    c.Publish("topic.t" + i, buffer);
                    Thread.Sleep(1);
                }
            }
            finally
            {
                foreach (var c in subscribers) c.Dispose();
                foreach (var c in publishers) c.Dispose();
            }
            Console.WriteLine("complete");
        }

    }
```

However, the problem is getting harder if I consider a multi-thread pattern, and I distribute the connections as one-per-thread. Under this context, it works decently with up a dozen of connections, and some delay somewhere. If you try to consider 100+ connections, simply NONE will succeed!
```
    class NatsOrigMultiThreadDemo
    {
        const int N = 1 * 1000;

        string[] servers = new string[]
        {
            //"nats://127.0.0.1:4222",
            "nats://localhost:4222"
        };


        public void Run()
        {
            var subscribers = Enumerable.Range(0, 100)
                .Select(_ => new Thread(this.Subscriber))
                .ToArray();

            var publishers = Enumerable.Range(0, 100)
                .Select(_ => new Thread(this.Publisher))
                .ToArray();

            for (int i = 0; i < subscribers.Length; i++)
            {
                subscribers[i].Start("topic" + i);
                //Thread.Sleep(3000);
            }

            //Thread.Sleep(5000);

            for (int i = 0; i < publishers.Length; i++)
            {
                //Thread.Sleep(500);
                publishers[i].Start("topic" + i);
            }

            foreach (var t in publishers) t.Join();
            foreach (var t in subscribers) t.Join();
        }


        private void Publisher(object state)
        {
            var subjectBase = (string)state;
            Console.WriteLine("Starting publisher " + subjectBase);

            Options opts = ConnectionFactory.GetDefaultOptions();
            opts.PingInterval = 30000;
            opts.Timeout = 3000;
            opts.Servers = servers;

            ConnectionFactory cf = new ConnectionFactory();
            using (IConnection c = cf.CreateConnection(opts))
            {
                var sw = Stopwatch.StartNew();
                for (int i = 0; i < N; i++)
                {
                    var jreq = new JObject();
                    jreq["n"] = i;
                    jreq["descr"] = "Creates a live connection";
                    jreq["dt"] = DateTime.Now;
                    byte[] buffer = Encoding.UTF8.GetBytes(jreq.ToString());
                    c.Publish(subjectBase + ".t" + i, buffer);
                    //Console.WriteLine("buffer-length=" + buffer.Length);
                    Thread.Sleep(1);
                }
                sw.Stop();
                Console.WriteLine($"Publisher end {subjectBase}; {sw.ElapsedMilliseconds}ms; {101 * N / 1000.0 / sw.ElapsedMilliseconds}MB/s");
                Thread.Sleep(3000);
            }
        }


        private void Subscriber(object state)
        {
            var topicBase = (string)state;
            Console.WriteLine("Starting subscriber " + topicBase);

            Options opts = ConnectionFactory.GetDefaultOptions();
            opts.PingInterval = 30000;
            opts.Timeout = 3000;
            opts.Servers = servers;

            ConnectionFactory cf = new ConnectionFactory();
            using (IConnection c = cf.CreateConnection(opts))
            {
                const int count = 1;
                Thread.Sleep(10000);
#if ASYNC
                var consumers = Enumerable.Range(0, count)
                    .Select(_ => ConsumerAsync.Create(c, topicBase + ".>"))
                    .ToArray();

                ManualResetEvent.WaitAll(consumers.Select(_ => _.Event).ToArray());
#endif

#if SYNC
                var cde = new CountdownEvent(count);
                for (int i = 0; i < count; i++)
                {
                    var cs = ConsumerSync.Create(c, topicBase + ".>", cde);
                    ThreadPool.QueueUserWorkItem(cs.Worker);
                }
                cde.Wait();
#endif
                Console.WriteLine("Subscriber end " + topicBase);
            }
        }


        private class ConsumerAsync
        {
            private ConsumerAsync() { }
            public static ConsumerAsync Create(IConnection conn, string topic)
            {
                var instance = new ConsumerAsync();
                instance._sa = conn.SubscribeAsync(topic, instance.Handler);
                return instance;
            }

            private IAsyncSubscription _sa;
            private int _n;
            public readonly ManualResetEvent Event = new ManualResetEvent(false);

            private void Handler(object sender, MsgHandlerEventArgs e)
            {
                var jres = JToken.Parse(Encoding.UTF8.GetString(e.Message.Data));
                var ix = (int)jres["n"];
                ++_n;
                if (_n == N) this.Event.Set();
            }
        }


        private class ConsumerSync
        {
            private ConsumerSync() { }
            public static ConsumerSync Create(IConnection conn, string topic, CountdownEvent cde)
            {
                var instance = new ConsumerSync();
                instance._sa = conn.SubscribeSync(topic);
                instance._cde = cde;
                return instance;
            }

            private ISyncSubscription _sa;
            private int _n;
            private CountdownEvent _cde;

            public void Worker(object state)
            {
                while (this._n < N)
                {
                    Msg m = this._sa.NextMessage();
                    var jres = JToken.Parse(Encoding.UTF8.GetString(m.Data));
                    var ix = (int)jres["n"];
                    ++_n;
                }
                this._cde.Signal();
            }
        }

    }
```

Really, I'm still scratching my head, but I wish to understand how to make more reliable yet fast the client with many connections.
Thanks in advance
Mario

**Comment by sixlettervariables** [comment]
You're basically exhausting the thread pool. On my 32 core workstation, things got unwieldy past the 500 thread mark...

Any reason why you need so many NATS connections? Within an application one connection can handle arbitrarily many publishers and subscribers.

Your second example fails because most of the publishers finish publishing before the subscribers are even ready to listen.

**Comment by ColinSullivan1** [comment]
Can you describe in more detail what the goals of your test are?  A NATS client typically needs only one connection (in some cases maybe a few more).

**Comment by highfield** [comment]
I agree that a single process could work with just one connection, but my goal is to simulate a larger environment with many processes. Let's say some hundreds of connections (i.e. concurrent processes) when the system will be publicly on. If you like, consider the broker in the cloud, and many different processes connected to.
So, the questions are still there: it should work fine with 200 processes (one connection each), but not with 200 threads?

@sixlettervariables the above multithread sample begins to fail even with 20-30 threads: I can't believe 30 threads are too many for a decent PC!...
Moreover, notice that the listeners do nothing: no subscription at all. Also the publishers begin to fire messages every millisecond or so: each one is around 200 bytes.
The problem is _NOT_ about the messages, rather the connections: the NATS client raises an exception because the first connection timeouts (hence all the others). I may take the timeout higher, let's say 15 seconds, and the things get better, but almost all the times some connection goes timeout.

In a short, why making a connection takes so long?

**Comment by ColinSullivan1** [comment]
I see.  I think this is a limitation of .NET due to default values with the thread pool.  Have you tried tuning your benchmark by setting `ThreadPool.SetMinThreads`? 

 If you are using .NET core 1.0, there is another way described [here](https://github.com/dotnet/corefx/issues/15990).

**Comment by ColinSullivan1** [comment]
To add to my comment, one hypothesis is that if you are running out of resources in the thread pool, additional tasks are taking longer to start once scheduled - which would include connection related tasks.  If the problem is mitigated by overriding the default number of threads, it'd help prove this out.

**Comment by highfield** [comment]
Will do: let me some time. At the moment, I'm busy on another project.
Many thanks in advance.

**Comment by sixlettervariables** [comment]
@highfield: regarding the increasing connection times, I saw two things during my testing:

- The internal NATS connection logic took between 5 and 20 ms, with some spikes to ~500ms.
- The dominating factor was waiting on the thread to start.

Internal to the NATS connection logic there is a bit more going on than just a new TCPClient plus a Connect. It waits for the server's INFO packet before continuing. That comprises the majority of the delay, but any local latency from resource exhaustion will increase that on the receive thread.

**Comment by sebastien-curutchet** [comment]
If the threadpool doesn't have enough available threads to do the connection, you have a timeout. It is because of this code client.ConnectAsync(s.url.Host, s.url.Port).Wait(TimeSpan.FromMilliseconds(timeoutMillis)). I have a similar issue on a stressed environment. I'm trying to implement the timeout by another way based on BeginConnect /EndConnect and TCPClient.Client.Poll method.

For now I did a workaround which avoids Timeout but doesn't resolve the delay to execute the Connect  method.

```csharp
public static bool Connect(this TcpClient client, string host, int port, int actionTimeoutInMs, int totalTimeoutInMs = int.MaxValue)
        {
            client.Client.SendTimeout = totalTimeoutInMs;
            ManualResetEvent mre = new ManualResetEvent(false);
            var task = Task.Run(() =>
            {
                mre.Set();
                client.Connect(host, port);
            });
            if (mre.WaitOne(totalTimeoutInMs))
            {
                if (!task.Wait(actionTimeoutInMs))
                {
                    return false;
                }
                return true;
            }
            return false;
        }
```

**Comment by ColinSullivan1** [comment]
@sebastien-curutchet , Thank you for for sharing the workaround.  This may be something we'll want to look into adding into the client.

**Comment by CITnDev** [comment]
Hello,
To complete my workaround this is another way to manage the timeout without using the ThreadPool : https://github.com/CITnDev/csharp-nats/commit/34c0e64402a484d8f755a3ba3a90c5af68ba4479

I have to do more tests in a real environment. But this change has an impact on the CreateConnection behavior. The UT TestConnectionMemoryLeak fails most of the time because in a few seconds the UT creates more than 10K connections : si I had to change the loop break condition. I don't have done deeper investigation to check if it is a client issue or a server one.

**Comment by ColinSullivan1** [comment]
This was resolved by #318.
